scala/private/dependency.bzl

```diff
@@ -0,0 +1,78 @@
+# This file contains all computations for what the dependency mode is
+# (i.e. transitive, plus-one, direct, etc)
+# and what/how dependency analysis is performed (unused deps, strict deps, etc).
+
+def new_dependency_info(
+        dependency_mode,
+        unused_deps_mode,
+        strict_deps_mode,
+        dependency_tracking_method):
+    is_strict_deps_on = strict_deps_mode != "off"
+
+    # Currently we do not support running both strict and unused deps at the same time
+    if is_strict_deps_on:
+        unused_deps_mode = "off"
+
+    is_unused_deps_on = unused_deps_mode != "off"
+
+    need_direct_jars = is_strict_deps_on or is_unused_deps_on
+    need_direct_targets = is_unused_deps_on
+
+    return struct(
+        dependency_mode = dependency_mode,
+        need_indirect_info = is_strict_deps_on,
+        need_direct_jars = need_direct_jars,
+        need_direct_targets = need_direct_targets,
+        need_direct_info = need_direct_jars or need_direct_targets,
+        dependency_tracking_method = "high-level",
+        unused_deps_mode = unused_deps_mode,
+        strict_deps_mode = strict_deps_mode,
+        use_analyzer = is_strict_deps_on or is_unused_deps_on,
+    )
+
+def dependency_info_for_addons(ctx):
+    return new_dependency_info(
+        dependency_mode = _dependency_mode_for_addons(ctx),
+        unused_deps_mode = "off",
+        strict_deps_mode = get_strict_deps_mode(ctx),
+        dependency_tracking_method = "high-level",
+    )
+
+def get_dependency_mode(ctx):
+    if get_is_strict_deps_on(ctx):
+        # all transitive dependencies are included
+        return "transitive"
+    elif not _is_plus_one_deps_off(ctx):
+        # dependencies and dependencies of dependencies are included
+        return "plus-one"
+    else:
+        # only explicitly-specified dependencies are included
+        return "direct"
+
+def _dependency_mode_for_addons(ctx):
```

Comment:
why don't they have +1? Just because that was the existing support? Something else?

Reply:
Yeah mainly that the existing code for scrooge/protobuf only had the two options of direct and strict. I'm actually not familiar with those rules at all so opted to preserve existing behavior.

Reply:
so maybe we should open an issue and add a comment about it here?
instead of trying to sugar coat it with `addons` let's say `_legacy_unclear_dependency_mode_for_protobuf_scrooge`. I know it might sound funny but it will make it clear for readers this isn't a well groomed piece but something we need to handle

Reply:
Added a comment and filed https://github.com/bazelbuild/rules_scala/issues/987